### ISSUE TITLE
What happens if the available range of stream identifiers is exhausted

### ISSUE THREAD

**Issue by Zsolt-LazarZsolt** [envoyproxy/envoy#19219]
**If you are reporting *any* crash or *any* potential security issue, *do not*
open an issue in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.**

*Title*: what is the behavior when https stream identifiers are exhausted.

*Description*:
In a setup with an upstream where h2 is set, but no max_stream_duration or max_requests_per_connection is set towards an upstream.

Is it possible to run into a problem where HTTP2 connections where the stream id is exhausted (more than 2^31).
What is the behavior of envoy as a client in the scenario?

Also the same question applies to downstream, does envoy implement the GOAWAY to handle these situations?

A related RFC chapter mentions this:

>    Stream identifiers cannot be reused.  Long-lived connections can
>    result in an endpoint exhausting the available range of stream
>    identifiers.  A client that is unable to establish a new stream
>    identifier can establish a new connection for new streams.  A server
>    that is unable to establish a new stream identifier can send a GOAWAY
>    frame so that the client is forced to open a new connection for new
>    streams.

We are using: 1.16 and up

[optional *Relevant Links*:]
https://datatracker.ietf.org/doc/html/rfc7540#section-5.1.1

Thank you for your answer in advance.

**Comment by Zsolt-LazarZsolt** [comment]
We have also seen a difference between documentation and code:
The code seem to be hard capping if not given:
https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_pool_base.cc#L94
https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_pool_base.cc#L172-L174

While the documentation does not mention any default behavior:
https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/protocol.proto#envoy-v3-api-field-config-core-v3-httpprotocoloptions-max-requests-per-connection

Do I miss something? 
Is there a maximum number of streams/requests by default?
Is it possible to exhaust all possible stream_ids?

**Comment by phlax** [comment]
cc @yanavlasov

**Comment by yanavlasov** [comment]
Unfortunately the [error is ignored](https://github.com/envoyproxy/envoy/blob/c01793eb32a58f122d9a54cca7a432565e8186c9/source/common/http/http2/codec_impl.cc#L547). I have marked this as bug.

**Comment by YaoZengzeng** [comment]
I'd like to fix this bug:)

**Comment by Zsolt-LazarZsolt** [comment]
Please also consider that the documentation about max_requests_per_connection does say that it is by default unlimited while in reality it is set to be 1>>29
This is the smallest of the problems, but the easiest to fix.

**Comment by YaoZengzeng** [comment]
As a client, if no max_requests_per_connection is configured, it would set the to `DEFAULT_MAX_STREAMS` which is 1 << 29 (ref: https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_pool_base.cc#L180).

Then it would become the `remaining_streams_` of `ActiveClient` (ref: https://github.com/envoyproxy/envoy/blob/main/source/common/conn_pool/conn_pool_base.cc#L600), every time before a new stream is created, it would be decremented by 1 (ref: https://github.com/envoyproxy/envoy/blob/main/source/common/conn_pool/conn_pool_base.cc#L180). 

When `remaining_streams_` becomes 0, the connection would enter DRAINING state (ref: https://github.com/envoyproxy/envoy/blob/main/source/common/conn_pool/conn_pool_base.cc#L184) and never be used.  So the default value of max_requests_per_connection for upstream is not **unlimited**. I would fix the incorrect description of the doc.

But for downstream, the default value of `max_requests_per_connection` is unlimited according to the code (ref: https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_manager_impl.cc#L295).

@yanavlasov @Zsolt-LazarZsolt WDYT?

**Comment by yanavlasov** [comment]
Makes sense. I think we should harmonize upstream and downstream behavior.

**Comment by Zsolt-LazarZsolt** [comment]
I would agree, a sync between upstream and downstream behavior would be great.
